mansur_adam_ndvi.py

```diff
@@ -0,0 +1,892 @@
+        if isinstance(band, str):
+            band = rxr.open_rasterio(band_path, masked=True).squeeze()
+
+        # Remove pixels outside valid range if provided
+        if valid_range:
+            mask = ((band < min(valid_range)) | (band > max(valid_range)))
+            band = band.where(~xr.where(mask, True, False))
+
+        return band
+    
+    
+    def _parse_product_id(self):
+        """Parses acquisition and processing info from Landsat filename
+        
+        Returns
+        -------
+        dict
+            Dictionary containing the parsed metadata
+        """
+        prod_id = os.path.basename(self._obj[0].path.values[()])
+        sat, corr, wrs, acq, proc, coll_num, coll_cat = prod_id.split("_")[:7]
+        
+        return {
+            "sensor": sat[1],
+            "satellite": sat[3:],
+            "wrs_path": wrs[:3],
+            "wrs_row": wrs[3:],
+            "acq_date": datetime.strptime(acq, "%Y%m%d"),
+            "proc_date": datetime.strptime(proc, "%Y%m%d"),
+            "coll_num": coll_num,
+            "coll_cat": coll_cat  
+        }  
+
+    
+    
+    
+def open_landsat(path, min_band=1, max_band=7):
+    """Combines Landsat TIFFs in a directory into an xarray
+    
+    Also adds two ad hoc attributes, band and path. Some methods, like
+    arr.where(), seem to lose track of these attributes in some cases.
+
+    Arguments
+    ---------
+    path : str
+        path to directory containing a set of Landsat TIFFs
+    min_band: int
+        minimum band to retrieve
+    max_band: int
+        maximum band to retrieve
+
+    Returns
+    -------
+    xarray DataArray
+         Array merging all matching TIFFs into one object
+
+    Raises
+    ------
+    IOError
+        If wrong number of bands found
+    """
+
+    # Read TIFFs along path matching selected bands, then add the QA band
+    paths = []
+    for fp in glob(os.path.join(path, f"*band[{min_band}-{max_band}]*.tif")):
+        paths.append(fp)
+    paths.sort()
+    paths.extend(glob(os.path.join(path, f"*pixel_qa*.tif")))
+    
+    # Raise an error if any of the expected bands are missing
+    expected = (max_band - min_band + 2)
+    if len(paths) != expected:
+        raise IOError(f"Wrong number of TIFFs found on {os.path.abspath(path)}"
+                      f" (found {len(paths)}, expected={expected})")
+
+    # Combine ordered list of TIFFs into an xarray
+    out_xr =[]
+    for fp in paths:
+        out_xr.append(rxr.open_rasterio(fp, masked=True).squeeze())
+
+        # Add min_band so that the band id matches original Landsat band
+        out_xr[-1]["band"] = len(out_xr) + min_band - 1
+        out_xr[-1]["path"] = os.path.realpath(fp)
+
+    # Rename the last band to "qa"
+    out_xr[-1]["band"] = "qa"
+
+    return xr.concat(out_xr, dim="band")
+
+
+
+def mean_ndvi_from_landsat(path, site_name, site_boundary=None):
+    """Calculates mean NDVI for a set of Landsat images
+    
```

Comment:
At what point do you decide a calculation is worth putting into its own function? This definitely keeps things cleaner in the main body of the code, but I didn't think to put this quick calculation into its own function.

Reply:
I think this function is close to doing too much, but it's really context dependent. For example, if you wanted to use the same complex series of functions across multiple workbooks, it might make sense to use a function. My (bad) habit is to overload functions and add too many branches (type checking, etc.)